pkg/pgmodel/series_set.go

```diff
@@ -0,0 +1,196 @@
+		return false
+	}
+	for !p.rows[p.rowIdx].Next() {
+		p.rows[p.rowIdx].Close()
+		p.rowIdx++
+		if p.rowIdx >= len(p.rows) {
+			return false
+		}
+	}
+	return true
+}
+
+// At returns the current storage.Series. It expects to get rows to contain
+// four arrays in binary format which it attempts to deserialize into specific types.
+// It also expects that the first two and second two arrays are the same length.
+func (p *pgxSeriesSet) At() storage.Series {
+	if p.rowIdx >= len(p.rows) {
+		return nil
+	}
+	arrays := p.rows[p.rowIdx].RawValues()
+
+	// Setting invalid data until we confirm that all data is valid.
+	p.err = errInvalidData
+
+	if len(arrays) != 4 {
+		return nil
+	}
+
+	ps := &pgxSeries{}
+
+	if err := ps.labelNames.DecodeBinary(nil, arrays[0]); err != nil {
+		return nil
+	}
+	if err := ps.labelValues.DecodeBinary(nil, arrays[1]); err != nil {
+		return nil
+	}
+
+	if len(ps.labelNames.Elements) != len(ps.labelValues.Elements) {
+		return nil
+	}
+
+	if err := ps.times.DecodeBinary(nil, arrays[2]); err != nil {
+		return nil
+	}
+
+	if err := ps.values.DecodeBinary(nil, arrays[3]); err != nil {
+		return nil
+	}
+
+	if len(ps.times.Elements) != len(ps.values.Elements) {
+		return nil
+	}
+
+	p.err = nil
+	return ps
+}
+
+// Err implements storage.SeriesSet.
+func (p *pgxSeriesSet) Err() error {
+	return p.err
+}
+
+// pgxSeries implements storage.Series.
+type pgxSeries struct {
+	labelNames  pgtype.TextArray
+	labelValues pgtype.TextArray
+	times       pgtype.TimestamptzArray
+	values      pgtype.Float8Array
+}
+
+// Labels returns the label names and values for the series.
+func (p *pgxSeries) Labels() labels.Labels {
+	ll := make(labels.Labels, len(p.labelNames.Elements))
+
+	for i := range ll {
+		ll[i].Name = p.labelNames.Elements[i].String
+		ll[i].Value = p.labelValues.Elements[i].String
+	}
+
+	sort.Sort(ll)
+
+	return ll
+}
+
+// Iterator returns a chunkenc.Iterator for iterating over series data.
+func (p *pgxSeries) Iterator() chunkenc.Iterator {
+	return newIterator(p.times, p.values)
+}
+
+// pgxSeriesIterator implements storage.SeriesIterator.
+type pgxSeriesIterator struct {
+	cur          int32
+	totalSamples int32
+	times        pgtype.TimestamptzArray
+	values       pgtype.Float8Array
+}
+
+// newIterator returns an iterator over the samples. It expects times and values to be the same length.
+func newIterator(times pgtype.TimestamptzArray, values pgtype.Float8Array) *pgxSeriesIterator {
+	return &pgxSeriesIterator{
+		cur:          -1,
+		totalSamples: int32(len(times.Elements)),
+		times:        times,
+		values:       values,
+	}
+}
+
+// Seek implements storage.SeriesIterator.
+func (p *pgxSeriesIterator) Seek(t int64) bool {
+	found := false
+	p.cur = 0
+
+	for !found && p.cur < p.totalSamples {
+		if p.getTs() >= t {
+			found = true
+			break
+		}
+		p.cur++
+	}
+
+	return found
+}
+
+func (p *pgxSeriesIterator) getTs() int64 {
+	v := p.times.Elements[p.cur]
+
+	if v.Status == pgtype.Null {
+		return 0
```

Comment:
should probably be a two-value return of (model.Time, bool). 0 is a valid time value so we shouldn't use this to indicate NULL.